siyavula/latex2image/utils.py

```diff
@@ -46,3 +50,91 @@ def copy_if_newer(src, dest):
         success = False
 
     return success
+
+
+def unescape(text):
+    '''Formats unicode characters correctly'''
+    def fixup(m):
+        text = m.group(0)
+        if text[:2] == "&#":
+            # character reference
+            try:
+                if text[:3] == "&#x":
+                    return unichr(int(text[3:-1], 16))
+                else:
+                    return unichr(int(text[2:-1]))
+            except ValueError:
+                pass
+        else:
+            # named entity
+            try:
+                text = unichr(htmlentitydefs.name2codepoint[text[1:-1]])
+            except KeyError:
+                pass
+        return text  # leave as is
+    return re.sub("&#?\w+;", fixup, text)
+
+
+def cleanup_code(code):
+    ''' Removes nested math delimiters of the form \( \) inside $ $ pairs'''
+    result = re.findall(r'\$(.*?)\$', code)
+    for snippet in result:
+        newsnippet = snippet.replace(r'\(', ' ').replace(r'\)', ' ')
+        code = code.replace(snippet, newsnippet)
+
+    code = code.strip()
+
+    # if align* or {align} in code we don't need delimiters
+    if (r'align*' in code) or ('{align}' in code):
+        if code.startswith(r'\('):
+            code = code[2:]
+        if code.endswith('\)'):
+            code = code[:-2]
+
+    # remove blank lines and lines that start with %
+    newcode = []
+    for line in code.split('\n'):
+        if line.strip().startswith('%'):
+            continue
+        if not line.strip():
+            continue
+        newcode.append(line)
+    code = '\n'.join(newcode)
+
+    code = repair_equations(code)
+
+    return code
+
+
+def unicode_replacements(latex):
+    '''Takes in latex and replaces specific unicode characters with latex symbols'''
+    unicode_operators = {
+        "\xe2\x88\x92": '-',
+        "\xc3\x97": r'\times',
+        "\xc2\xb7": r'\cdot ',
+    }
+    unicode_superscripts = {
+        "\xc2\xb0": r'\text{$^\circ$}',
+        "\xe2\x81\xbb\xc2\xb9": r'^{-1}',
+        "\xc2\xb2": r'^{2}',
+        "\xc2\xb3": r'^{3}',
+        "\xe2\x84\x83": r'^{\circ}C',
+    }
+    unicode_punctation_spacing = {
+        "\xc2\xa0": ' ',
+    }
+    unicode_symbols = {
+        "\xce\xa9": r'\ensuremath{\Omega}',
+        "\xe2\x82\xac": r'\euro',
+    }
+
+    for key in unicode_operators:
```

Comment:
Remember to ask yourself when you see the same pattern repeated again and again... what if the bits that are the same were separated from those that are different (in this case, the method of traversing the dictionaries is the same, but the dictionaries themselves are different).